pytd/writer.py

```diff
@@ -104,14 +116,16 @@ def _cast_dtypes(dataframe, inplace=True, keep_list=False):
                 df[column] = df[column].apply(_convert_nullable_str, args=(str,))
             else:
                 t = str
+        elif kind == "b":
+            # Bulk Import API internally handles boolean string as a boolean type,
+            # and hence "True" ("False") will be stored as "true" ("false"). Align
+            # to lower case here.
+            t = object
+            df[column] = df[column].apply(
+                _convert_nullable_str, args=(bool,), lower=True
+            )
```

Comment:
What this clause does seems to be duplicated with the one captured above: `df[column].apply(_isinstance_or_null, args=(bool,)).all()`. 

Can we just merge this condition with `kind == "O"`?

```py
elif kind in ("b", "O"):
    t = object
    if df[column].apply(_isinstance_or_null, args=((list, np.ndarray),)).all():
        if keep_list:
            df[column] = df[column].apply(_to_list)
        else:
            df[column] = df[column].apply(
                _convert_nullable_str, args=((list, np.ndarray),)
            )
    elif df[column].apply(_isinstance_or_null, args=(bool,)).all():
        df[column] = df[column].apply(
            _convert_nullable_str, args=(bool,), lower=True
        )
    elif df[column].apply(_isinstance_or_null, args=(str,)).all():
        df[column] = df[column].apply(_convert_nullable_str, args=(str,))
    else:
        t = str
```

Reply:
Nice catch 👍 ef4ce92